[DOCU-2306] Websocket validator functionality plugins (#4150)
Kong/docs.konghq.com: src/gateway/reference/proxy.md

Upgrade: WebSocket
```

This makes {{site.base_gateway}} forward the `Connection` and `Upgrade` headers to your
upstream service, instead of dismissing them due to the hop-by-hop nature of a
standard HTTP proxy.

### WebSocket proxy modes

There are two methods for proxying WebSocket traffic in {{site.base_gateway}}:

* HTTP(S) services and routes
* WS(S) services and routes

{% navtabs %}
{% navtab HTTP(S) services and routes %}

Services and routes using the `http` and `https` protocols are fully capable of
handling WebSocket connections with no special configuration. With this method,
WebSocket sessions behave identically to regular HTTP requests, and all of the
request/response data is treated as an opaque stream of bytes.

```yaml
services:
  - name: my-http-websocket-service
    protocol: http
    host: 1.2.3.4
    port: 80
    path: /
    routes:
      - name: my-http-websocket-route
        protocols:
          - http
          - https
```

{% endnavtab %}

{% navtab WS(S) services and routes %}

> {:.badge .enterprise}

Reviewer: @cloudjumpercat when you have a sec, could you take a look and see if there's a better place for this enterprise badge? Right now it's not great:

![image](https://user-images.githubusercontent.com/3277009/181599685-069cfdd3-d116-4a48-8664-65153216b27e.png)

https://deploy-preview-4150--kongdocs.netlify.app/gateway/latest/reference/proxy/#proxy-websocket-traffic

My intent was to have the badge _in_ the navbar title for this (`WS(S) Services and Routes {{ENTERPRISE}}`).

Reviewer: I'm honestly not 100% sold on the navbar for this section anyways, since it makes the page less friendly to `ctrl+f` (which is a shame for something in the "reference" category). Maybe I'll just go back to plain markdown sections.

Reviewer: I decided to remove the navbar from this section.

Upgrade: WebSocket
```

This makes {{site.base_gateway}} forward the `Connection` and `Upgrade` headers to your
upstream service, instead of dismissing them due to the hop-by-hop nature of a
standard HTTP proxy.

### WebSocket Proxy Modes

There are two methods for proxying WebSocket traffic in {{site.base_gateway}}:

* HTTP(S) services and routes
* WS(S) services and routes

{% navtabs %}
{% navtab HTTP(S) Services and Routes %}

Services and routes using the `http` and `https` protocols are fully capable of
handling WebSocket connections with no special configuration. With this method,

> WebSocket sessions behave identically to regular HTTP requests, and all of the

Reviewer: This is a compound sentence, so the comma needs to stay (feel free to suggest different wording if you think there's a better way to state this though).

request/response data is treated as an opaque stream of bytes.

```yaml
services:
  - name: my-http-websocket-service
    protocol: http
    host: 1.2.3.4
    port: 80
    path: /
    routes:
      - name: my-http-websocket-route
        protocols:
          - http
          - https
```

{% endnavtab %}

{% navtab WS(S) Services and Routes %}
{:.badge .enterprise}

In addition to HTTP services and routes, {{site.ee_product_name}} includes
the `ws` (WebSocket-over-http) and `wss` (WebSocket-over-https) options for
service `protocol` and route `protocols`. In contrast to `http`/`https`, `ws`
and `wss` services have full control over the underlying WebSocket connection,
meaning they can use WebSocket plugins and the [WebSocket PDK](LINK_TO_PDK) to
perform business logic on a per-message basis (message validation, accounting,
rate-limiting, etc).

```yaml
services:
  - name: my-dedicated-websocket-service
    protocol: ws
    host: 1.2.3.4
    port: 80
    path: /
    routes:
      - name: my-dedicated-websocket-route
        protocols:
          - ws
          - wss
```

{% endnavtab %}
{% endnavtabs %}

{:.note}

> > _A note about performance:_
> >
> > Decoding and encoding WebSocket messages comes with a non-zero amount of
> > performance overhead when compared to the protocol-agnostic behavior of
> > `http(s)` services. If your API does not need the extra capabilities
> > provided by a `ws(s)` service, it is generally recommended to use an `http(s)`
> > service instead.

Author: ```suggestion
> **Note**:
> Decoding and encoding WebSocket messages comes with a non-zero amount of
> performance overhead when compared with the protocol-agnostic behavior of
> `http(s)` services. If your API does not need the extra capabilities
> provided by a `ws(s)` service, it's recommended that you use an `http(s)`
> service instead.
```

### WebSocket and TLS

Regardless of which service/route protocols are in use (`http(s)` or `ws(s)`),
{{site.base_gateway}} will accept plain and TLS WebSocket connections on its
respective `http` and `https` ports. To enforce TLS connections from clients,
set the `protocols` property of the [route][route-entity] to `https` or `wss`
only.

When setting up the [service][service-entity] to point to your upstream
WebSocket service, you should carefully pick the protocol you want to use
between {{site.base_gateway}} and the upstream.

> If you want to use TLS, your upstream WebSocket service must be defined using
> the `https` (or `wss`) protocol in the service `protocol` property, and the
> proper port (usually 443). To connect without TLS, then the `http` (or `ws`)

Author: ```suggestion
If you want to use TLS, your upstream WebSocket service must be defined using
the `https` (or `wss`) protocol in the service `protocol` property and the
proper port (usually 443). To connect without TLS, then the `http` (or `ws`)
```

upstream service, instead of dismissing them due to the hop-by-hop nature of a
standard HTTP proxy.

> If you are using either of the [WebSocket Size Limit](pluginlink) or the [WebSocket Validator](pluginlink) plugins, this may slow the performance of the proxy WebSocket traffic.

Reviewer: @cloudjumpercat I need to fill out this section some more to explain the differences between service/route protocols (http/https/ws/wss) for WebSocket traffic and when one might use one or the other, _including_ a call-out that ws/wss services are an EE-only feature.

I'm holding off on adding that for the moment in case you are still doing some more branch maintenance that would require a rebase/force-push. Can you @ me when you're ready for me to add more changes to the branch?

Author: @flrgh Okay, I should be done with editing branches for now (just had one comment on a placeholder value). I'll probably revisit this tomorrow to see if I can add to the PDK/Admin API doc.
